### PR TITLE
misspell: add extra-words

### PR DESCRIPTION
**Pull request by ldez**
Allows to add custom typo/corrections.

### CODE REVIEW

File: pkg/config/linters_settings.go [hunk 1]
```diff
@@ -663,12 +663,18 @@ type MalignedSettings struct {
 }
 
 type MisspellSettings struct {
-	Mode   string `mapstructure:"mode"`
-	Locale string
+	Mode       string `mapstructure:"mode"`
+	Locale     string
```
Comment (Antonboom) [inline pkg/config/linters_settings.go:667]:
```suggestion
	Locale     string `mapstructure:"locale"`
```

File: pkg/config/linters_settings.go [hunk 2]
```diff
@@ -663,12 +663,18 @@ type MalignedSettings struct {
...
+	ExtraWords []MisspellExtraWords `mapstructure:"extra-words"`
 	// TODO(ldez): v2 the option must be renamed to `IgnoredRules`.
 	IgnoreWords []string `mapstructure:"ignore-words"`
 }
 
+type MisspellExtraWords struct {
+	Typo       string
```
Comment (Antonboom) [inline pkg/config/linters_settings.go:674]:
```suggestion
	Typo       string `mapstructure:"typo"`
```

File: pkg/config/linters_settings.go [hunk 3]
```diff
@@ -663,12 +663,18 @@ type MalignedSettings struct {
...
+	Correction string
```
Comment (Antonboom) [inline pkg/config/linters_settings.go:675]:
```suggestion
	Correction string `mapstructure:"correction"`
```

File: pkg/golinters/misspell.go [hunk 4]
```diff
@@ -153,3 +159,24 @@ func runMisspellOnFile(lintCtx *linter.Context, filename string, replacer *missp
 
 	return res, nil
 }
+
+func appendExtraWords(replacer *misspell.Replacer, extraWords []config.MisspellExtraWords) error {
```
Comment (Antonboom) [inline pkg/golinters/misspell.go:163]: no tests 😢

File: pkg/golinters/misspell.go [hunk 5]
```diff
@@ -95,6 +96,11 @@ func createMisspellReplacer(settings *config.MisspellSettings) (*misspell.Replac
 		return nil, fmt.Errorf("unknown locale: %q", settings.Locale)
 	}
 
+	err := appendExtraWords(replacer, settings.ExtraWords)
+	if err != nil {
+		return nil, err
```
Comment (Antonboom) [inline pkg/golinters/misspell.go:101]:
```suggestion
		return nil, fmt.Errorf("process extra words: %v": err)
```

File: pkg/golinters/misspell.go [hunk 6]
```diff
@@ -153,3 +159,24 @@ func runMisspellOnFile(lintCtx *linter.Context, filename string, replacer *missp
 
 	return res, nil
 }
+
+func appendExtraWords(replacer *misspell.Replacer, extraWords []config.MisspellExtraWords) error {
+	var extra []string
```
Comment (Antonboom) [inline pkg/golinters/misspell.go:164]: preallocation?

File: pkg/golinters/misspell.go [hunk 7]
```diff
@@ -153,3 +159,24 @@ func runMisspellOnFile(lintCtx *linter.Context, filename string, replacer *missp
...
+
+	for _, word := range extraWords {
+		if word.Typo == "" || strings.ContainsFunc(word.Typo, func(r rune) bool { return !unicode.IsLetter(r) }) {
```
Comment (Antonboom) [inline pkg/golinters/misspell.go:167]: please, separate the "empty value" and "not letters" errors

File: pkg/golinters/misspell.go [hunk 8]
```diff
@@ -153,3 +159,24 @@ func runMisspellOnFile(lintCtx *linter.Context, filename string, replacer *missp
...
+			return fmt.Errorf("the word in the 'typo' field should only contain letters")
+		}
+		if word.Correction == "" || strings.ContainsFunc(word.Correction, func(r rune) bool { return !unicode.IsLetter(r) }) {
```
Comment (Antonboom) [inline pkg/golinters/misspell.go:170]: DRY, let's to make the helper function?

**Review by Antonboom — CHANGES_REQUESTED** [review]

File: pkg/golinters/misspell.go (hunk 6 above)
Comment (ldez) [reply pkg/golinters/misspell.go:164]: preallocation is globally useless unless you have performance issues.

File: pkg/golinters/misspell.go [hunk 9]
```diff
@@ -153,3 +159,30 @@ func runMisspellOnFile(lintCtx *linter.Context, filename string, replacer *missp
...
+		}
+
+		if strings.ContainsFunc(word.Typo, func(r rune) bool { return !unicode.IsLetter(r) }) {
+			return fmt.Errorf("the word %q in the 'typo' field should only contain letters", word.Typo)
+		}
+		if strings.ContainsFunc(word.Correction, func(r rune) bool { return !unicode.IsLetter(r) }) {
+			return fmt.Errorf("the word %q in the 'correction' field should only contain letters", word.Correction)
+		}
+
+		extra = append(extra, strings.ToLower(word.Typo), strings.ToLower(word.Correction))
```
Comment (Antonboom) [inline pkg/golinters/misspell.go:182]: cosmetic: add to config some comment about case insensitivity?

File: test/testdata/misspell_custom.go [hunk 10]
```diff
@@ -0,0 +1,10 @@
+//golangcitest:args -Emisspell
```
Comment (Antonboom) [inline test/testdata/misspell_custom.go:1]: ❤️

File: pkg/golinters/misspell_test.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+	extraWords := []config.MisspellExtraWords{
+		{
+			Typo:       "iff",
+			Correction: "if",
+		},
+		{
+			Typo:       "cancelation",
+			Correction: "cancellation",
+		},
+	}
```
Comment (Antonboom) [inline pkg/golinters/misspell_test.go:23]:
```suggestion
		{
			Typo:       "oCcUreD",
			Correction: "OcCuRrEd",
		},
	}
```

**Review by Antonboom — APPROVED** [review]

File: test/testdata/configs/misspell_custom.yml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+linters-settings:
+  misspell:
+    extra-words:
+      - typo: "iff"
```
Comment (alexandear) [inline test/testdata/configs/misspell_custom.yml:4]: I'm not sure if `iff` is a good example of a typo. "iff" can stand for ["If and only if"](https://en.wikipedia.org/wiki/If_and_only_if).

Some time ago, I created a [PR](https://github.com/dominikh/go-tools/pull/1375) to fix `iff` to `if`, but it was rejected.

Comment (ldez) [reply test/testdata/configs/misspell_custom.yml:4]: https://go.dev/wiki/Spelling

**Pull request merged**